test(inter-protocol): snapshot storage/RPC API for auction, vaultFactory, reserve, psm, committees

dckc opened the pull request:
refs: #6111

in particular: https://github.com/Agoric/agoric-sdk/issues/6111#issuecomment-1481366083

## Description

Flesh out the "Reading data off-chain" section of the README with snapshots of chainStorage data produced by the inter protocol contracts. 

**NOTE:** The README says `auction.books.book0`, but what shows up in the snapshot is just `auction.books0`. Is that by design?

If feedback on this draft shows that we do want this sort of test...

DRAFT TODO:
  - [x] address copy-pasta from on-exit test

### Security / Scaling Considerations

none, AFAICT.

### Documentation Considerations

adds reliable docs.

rendered: [Reading data off-chain](https://github.com/Agoric/agoric-sdk/tree/dc-auction-rpc-snapshot/packages/inter-protocol#reading-data-off-chain)

for example:
 - [snapshots/test-auctionContract.js.md](https://github.com/Agoric/agoric-sdk/blob/dc-auction-rpc-snapshot/packages/inter-protocol/test/auction/snapshots/test-auctionContract.js.md)

I think consumers look for "RPC API". I wonder if we should tweak the "Reading data off-chain" section heading to include that.

### Testing Considerations

This adds a bit of test maintenance. As usual with ava snapshot tests: `yarn test test/auction/test-auctionContract.js -m '*snap*' --update-snapshots`

turadg reviewed (commented):
Great idea.

> NOTE: The README says auction.books.book0, but what shows up in the snapshot is just auction.books0. Is that by design?

I missed that in review. Good thing for snapshot tests! @Chris-Hibbert could you push a commit to correct that here?

Chris-Hibbert commented:
> The README says auction.books.book0, but what shows up in the snapshot is just auction.books0. Is that by design?

I was guessing, I'll admit. I'm definitely okay with it the way it is. If there's a reason for it to be otherwise, it should be simple to fix.

> could you push a commit to correct that here?

Done  3cc073a25

Chris-Hibbert commented on packages/inter-protocol/test/auction/test-auctionContract.js:
```diff
@@ -1250,3 +1253,102 @@ test('bid zero', async t => {
 });
+
+test('auctioneer chainStorage RPC API snapshot', async t => {
+  //#region copied from on-exit test
```
I'm not a fan of #region comments.

dckc replied:
agreed that this one, at least, doesn't carry its weight. As noted above, I aim to "DRY out copy-pasta from on-exit test".

dckc replied:
I didn't find a straightforward way to DRY out the common stuff; I changed the `#region` comments into a function.

dckc replied:
it occurs to me that, rather than repeating a bunch of stuff from another test, I can just add the `t.snapshot()` into that test.

dckc commented on packages/inter-protocol/README.md:
```diff
@@ -48,12 +48,10 @@ The canonical keys (under `published`) are as follows. Non-terminal nodes could
-        - `books`
-          - `book0`
-            - `data`
+        - `book0`
```
so we want `book0`, `book1` etc. under `auction` after all?

I'm standing by to see if by "correct this" @turadg meant to change the published structure to include a `book` or `books` node rather than tweaking this README.

turadg replied:
each commit ends up separately in master so a bugfix commit is reasonable to me to include in any PR that has the appropriate reviewers. if you want to do the work in a different PR up to you.

as an issue to prioritize, https://github.com/Agoric/agoric-sdk/issues/6111 encompasses the requirement. it's prioritized as a blocker for Vaults Release.

Chris-Hibbert replied:
I tried integrating the change with the existing commit stack, but the changes to the README are in a different commit than where I'd put the code changes. When I tried it, it turned into a mess, because several later commits also edit the same section of the README.  It'll be far simpler as a follow-on.

dckc commented on packages/governance/test/unitTests/snapshots/test-committee.js.md:
```diff
@@ -0,0 +1,62 @@
+          },
+          electionType: 'survey',
+          issue: {
+            text: 'why3',
```
This lacks detail on parameter change issues. I think developers of dapp-gov, i.e. us, are the most interested consumers of that API, and we already figured it out.

dckc commented on packages/inter-protocol/README.md:
````diff
@@ -30,6 +30,8 @@ To maintain that the keys of vaults to liquidate are stable requires that its ke
 
 VaultFactory publishes data using StoredPublishKit which tees writes to off-chain storage. These can then be followed off-chain like so,
 ```js
+import { makeFollower } from '@agoric/casting';
````
Where does `makeDefaultLeader` come from? I don't see it in `@agoric/casting`.

dckc commented on packages/inter-protocol/README.md:
```diff
@@ -40,32 +42,35 @@ VaultFactory publishes data using StoredPublishKit which tees writes to off-chai
+      - `governance`
+      - `metrics`
+    - `priceFeed` - [snapshot of details](./test/price/snapshots/test-fluxAggregator.js.md)
+      - `${inputBrand}-${outputBrand}_price_feed`
```
this formerly had a `_` as in `ATOM_USD` when snapshotting shows it's `ATOM-USD`.

dckc commented on packages/inter-protocol/test/price/snapshots/test-fluxAggregator.js.md:
```diff
@@ -0,0 +1,45 @@
+        'published.priceAggregator.LINK-USD_price_feed.latestRound',
+        {
+          roundId: 3n,
+          startedAt: 5n,
```
i suppose this will change to a time record in #5799 so that we can keep from exposing the whole timer below

dckc commented:
The technique I eventually found for the auction contract was so straightforward that I applied it to the other contracts as well (except stakeFactory).

PTAL.

dckc commented:
## Basis points vs. Ratios

We seem to publish basis points in some cases. I wonder if the econ-gov UI is set up for that.

```js
        LowestRate: {
          type: 'nat',
          value: 4500n,
        },
```

Chris-Hibbert reviewed (approved):
I'm not an expert on chainStorage, but this looks fine to me.

turadg commented on packages/governance/test/unitTests/test-committee.js:
```diff
@@ -134,7 +134,28 @@ test('committee-open question:one', async t => {
 });
 
-test('committee-open question:mixed', async t => {
+const snapshotStorage = async (t, storage) => {
```
types please
```suggestion
const snapshotStorage = async (
  /** @type {import('ava').ExecutionContext<unknown>} */ t,
  /** @type {import('@agoric/internal/src/storage-test-utils.js').MockChainStorageRoot} */ storage,
) => {
```

turadg commented on packages/inter-protocol/test/auction/test-auctionContract.js:
```diff
@@ -906,6 +924,8 @@ test.serial('onDeadline exit', async t => {
     currentPriceLevel: { numerator: { value: 9_350_000_000_000n } },
   });
 
+  await snapshotStorage(t, driver.mockChainStorage);
```
each test is defining its own `snapshotStorage` but they follow the same pattern. please DRY it out. I think something like,
```suggestion
  await snapshotStorage(t, driver.mockChainStorage, {'mockChainStorageRoot.thisAuction': 'published.auction'}, `Under "published", the "auction" node is delegated to
the auctioneer contract, which publishes data as follows.
See also board marshalling conventions (_to appear_).
` );
```

or maybe more to the point: `documentStorageKeys` since the `note` and the rewriting to `published:` is not really about testing.

turadg commented on packages/inter-protocol/test/psm/test-psm.js:
```diff
@@ -182,7 +182,10 @@ async function makePsmDriver(t, customTerms) {
       feeMintAccess,
       initialPoserInvitation,
-      storageNode: mockChainStorage.makeChildNode('thisPsm'),
+      storageNode: mockChainStorage
+        .makeChildNode('psm')
+        .makeChildNode('IST')
+        .makeChildNode('AUSD'),
```
this looks like logic that is meant to be under test but is part of the driver.

consider a function to make a storage path for a PSM which can be re-used in startPSM:
https://github.com/Agoric/agoric-sdk/blob/50c8e2681fe8bd668278fde3e97c0333e8803cfd/packages/inter-protocol/src/proposals/startPSM.js#L129-L133

dckc commented on packages/governance/test/unitTests/test-committee.js:
```diff
@@ -134,7 +134,28 @@ test('committee-open question:one', async t => {
+    storage.getBody(key),
+  ]);
+
+  const note = `Under "published", the "committees.Economic_Committee" node is delegated to
+a committee contract, which publishes data as follows.
```
It's supposed to be illustrative. Ideally, it would illustrate all branches of any union types. I'm not sure how to be precise about what the reader can count without putting so much text in the way that it's distracting..

> A plausible instance?

That's accurate and concise, if not complete. I suppose I can go with that.

turadg commented on packages/governance/test/unitTests/test-committee.js:
```diff
@@ -224,4 +225,12 @@ test('committee-open question:mixed', async t => {
+  const doc = {
+    node: 'committees.Economic_Committee',
+    owner: 'a committee contract',
+    pattern: 'mockChainStorageRoot.thisElectorate.',
+    replacement: 'published.committees.Economic_Committee.',
```
these would read more easily as a map, even if it has just one item.
```suggestion
    rekey: {'mockChainStorageRoot.thisElectorate': published.committees.Economic_Committee`},
```

turadg reviewed (approved):
Nicely done

dckc commented:
A couple storage keys in the vaults snapshot flip-flopped. I'm not sure how stable they were. I sorted them all an regenerated the snapshots.

Also, I squashed all the inter-protocol snapshot commits together. And moved all the governance commits before inter-protocol.
